src/radical/pilot/agent/scheduler/continuous.py

```diff
@@ -70,258 +76,387 @@ def finalize_child(self):
+
+        We don't care about continuity within a single node - cores `[1,5]` are
+        assumed to be as close together as cores `[1,2]`.
+
+        When `chunk` is set, only sets of exactly that size (or multiples
+        thereof) are considered valid allocations.  The use case is OpenMP
+        support, where each process is expected to create a certain number of
+        threads, which thus must have slots on the same node available.
+
+        NOTE: chunking is only applied to cores at this point.
+
+        When `partial` is set to `True`, we also accept less cores and gpus then
+        requested (but the call will never return more than requested).
+        '''
+
+
+        cores = list()
+        gpus  = list()
+
+        free_cores = node['cores'].count(rpc.FREE)
+        free_gpus  = node['gpus' ].count(rpc.FREE)
+
+        # first count the free cores/gpus, as that is way quicker than
+        # actually finding the core IDs.
+        if partial:
+            # For partial requests the check simpliefies: we just check if we 
+            # have either, some cores *or* gpus, to serve the request
+            if  (requested_cores and not free_cores) and \
+                (requested_gpus  and not free_gpus )     :
+                # wa can't serve either request
+                return [], []
 
-        if not 'task_slots' in opaque_slots:
-            raise RuntimeError('insufficient information to release slots via %s: %s' \
-                    % (self.uid, opaque_slots))
+        else:
+            # non-partial requests (ie. full requests): check if we can serve
+            # both, requested cores *and* gpus.
+            if  requested_cores > free_cores or \
+                requested_gpus  > free_gpus     :
+                return [], []
 
-        self._change_slot_states(opaque_slots['task_slots'], rpc.FREE)
 
+        # we can serve the partial or full request - alloc the chunks we need
+        # FIXME: chunk gpus, too?
+        alloc_cores = min(requested_cores, free_cores) / chunk * chunk
+        alloc_gpus  = min(requested_gpus , free_gpus )
 
-    # --------------------------------------------------------------------------
-    #
-    # Find a needle (continuous sub-list) in a haystack (list)
-    #
-    def _find_sublist(self, haystack, needle):
-        n = len(needle)
-        # Find all matches (returns list of False and True for every position)
-        hits = [(needle == haystack[i:i+n]) for i in xrange(len(haystack)-n+1)]
-        try:
-            # Grab the first occurrence
-            index = hits.index(True)
-        except ValueError:
-            index = None
+        # now dig out the core and gpu IDs.
+        for idx,state in enumerate(node['cores']):
+            if alloc_cores == len(cores):
+                break
+            if state == rpc.FREE:
+                cores.append(idx)
 
-        return index
+        for idx,state in enumerate(node['gpus']):
```

Comment:
Is there a reason for this for loop to run every time? That goes for the cores loop. Wouldn't a check make sense, to avoid the time consumed in the loops?

Reply:
Yes, that would make sense.  But I would like to keep the optimization parts out of this PR, and address them separately.